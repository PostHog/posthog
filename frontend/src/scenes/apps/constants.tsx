import { AppMetricsTab } from '~/types'

interface Description {
    successes: string
    successes_tooltip?: React.ReactNode | string

    successes_on_retry?: string
    successes_on_retry_tooltip?: React.ReactNode | string

    failures: string
    failures_tooltip?: React.ReactNode | string
}

export const DescriptionColumns: Record<string, Description> = {
    [AppMetricsTab.ProcessEvent]: {
        successes: 'Events processed',
        successes_tooltip: (
            <>
                These events were successfully processed and transformed by the <i>processEvent</i> app method.
            </>
        ),
        failures: 'Failed events',
        failures_tooltip: (
            <>
                These events had errors when being processed by the <i>processEvent</i> app method, but were still
                ingested.
            </>
        ),
    },
    [AppMetricsTab.OnEvent]: {
        successes: 'Events processed',
        successes_tooltip: (
            <>
                These events were successfully processed by the <i>onEvent</i> app method on the first try.
            </>
        ),
        successes_on_retry: 'Events processed on retry',
        successes_on_retry_tooltip: (
            <>
                These events were successfully processed by the <i>onEvent</i> app method after being retried.
            </>
        ),
        failures: 'Failed events',
        failures_tooltip: (
            <>
                These events had errors when being processed by the <i>onEvent</i> app method.
            </>
        ),
    },
    [AppMetricsTab.ComposeWebhook]: {
        successes: 'Events processed',
        successes_tooltip: (
            <>
                These events were successfully processed by the <i>composeWebhook</i> app method on the first try.
            </>
        ),
        successes_on_retry: 'Events processed on retry',
        successes_on_retry_tooltip: (
            <>
                These events were successfully processed by the <i>composeWebhook</i> app method after being retried.
            </>
        ),
        failures: 'Failed events',
        failures_tooltip: (
            <>
                These events had errors when being processed by the <i>composeWebhook</i> app method.
            </>
        ),
    },
    [AppMetricsTab.ExportEvents]: {
        successes: 'Events delivered',
        successes_tooltip: (
            <>These events were successfully delivered to the configured destination on the first try.</>
        ),
        successes_on_retry: 'Events delivered on retry',
        successes_on_retry_tooltip: (
            <>These events were successfully delivered to the configured destination after being retried.</>
        ),
        failures: 'Failed events',
        failures_tooltip: <>These events failed to be delivered to the configured destination due to errors.</>,
    },
    [AppMetricsTab.HistoricalExports]: {
        successes: 'Events delivered',
        successes_tooltip: (
            <>These events were successfully delivered to the configured destination on the first try.</>
        ),
        successes_on_retry: 'Events delivered on retry',
        successes_on_retry_tooltip: (
            <>These events were successfully delivered to the configured destination after being retried.</>
        ),
        failures: 'Failed events',
        failures_tooltip: <>These events failed to be delivered to the configured destination due to errors.</>,
    },
    [AppMetricsTab.ScheduledTask]: {
        successes: 'Successes',
        successes_tooltip: (
            <>
                How many times <i>runEveryMinute</i>, <i>runEveryHour</i> or <i>runEveryDay</i> was called successfully.
            </>
        ),
        failures: 'Failures',
        failures_tooltip: (
            <>
                How many times <i>runEveryMinute</i>, <i>runEveryHour</i> or <i>runEveryDay</i> failed due to errors.
            </>
        ),
    },
}
