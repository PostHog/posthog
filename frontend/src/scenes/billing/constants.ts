export const USAGE_TYPES = [
    { label: 'Events', value: 'event_count_in_period' },
    { label: 'Recordings', value: 'recording_count_in_period' },
    { label: 'Mobile Recordings', value: 'mobile_recording_count_in_period' },
    { label: 'Feature Flags', value: 'billable_feature_flag_requests_count_in_period' },
    { label: 'Exceptions', value: 'exceptions_captured_in_period' },
    { label: 'Rows Synced', value: 'rows_synced_in_period' },
    { label: 'Persons', value: 'enhanced_persons_event_count_in_period' },
    { label: 'Survey Responses', value: 'survey_responses_count_in_period' },
] as const

export type UsageTypeOption = (typeof USAGE_TYPES)[number]
export type UsageTypeValue = UsageTypeOption['value']

export const ALL_USAGE_TYPES: UsageTypeValue[] = USAGE_TYPES.map((opt) => opt.value)
