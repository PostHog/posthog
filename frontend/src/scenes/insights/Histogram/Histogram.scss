@import '../../../styles/breakpoints';

.dashboard-wrapper {
    width: 95%;
    height: 90%;
    margin: auto;
}

.histogram-container {
    display: flex;
    width: 100%;

    svg {
        margin: auto;

        g#bars {
            fill: var(--funnel-alt);
        }

        g#y-gridlines {
            g.tick:not(:first-child) {
                color: var(--funnel-axis);
            }

            .domain,
            g.tick:nth-of-type(1) {
                display: none; // hide axis line
            }
        }

        g#y-axis {
            .domain {
                stroke: var(--funnel-axis);
            }
        }

        g#x-axis {
            .domain {
                stroke: var(--funnel-axis-zero);
            }
        }

        g#x-axis,
        g#y-axis {
            text {
                font-size: 13px;
                fill: var(--funnel-axis-zero);

                @media (max-width: $md) {
                    font-size: 11px;
                }
            }
        }

        g#labels {
            text.bar-label {
                font-size: 14px;
                font-weight: bold;
                fill: var(--text-light);
                &.outside {
                    fill: var(--funnel-alt);
                }
            }
        }
    }
}
