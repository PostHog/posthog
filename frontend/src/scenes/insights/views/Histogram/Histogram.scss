@import '../../../../styles/mixins';

.dashboard-wrapper {
    width: 100%;
    height: calc(100% - 2rem);
}

.histogram-container {
    display: flex;

    svg {
        margin-right: 1rem;
        margin-left: 1rem;

        /*
         * Axis
         */
        g#y-axis {
            .domain {
                stroke: var(--funnel-axis);
            }
        }

        g#x-axis {
            .domain {
                stroke: var(--funnel-axis);
            }

            .tick {
                color: var(--funnel-axis);
            }
        }

        g#x-axis,
        g#y-axis {
            text {
                font-family: var(--font-sans);
                font-size: 0.75rem;
                font-weight: 500;
                fill: var(--default);
            }
        }

        /*
         * Grid
         */
        g#y-gridlines {
            stroke-dasharray: 4, 2;

            g.tick:not(:first-child) {
                color: var(--funnel-grid);
            }

            .domain,
            g.tick:nth-of-type(1) {
                display: none; // hide axis line
            }
        }

        /*
         * Bars
         */
        g#bars {
            fill: var(--primary-3000);
        }

        g#labels {
            text.bar-label {
                font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;

                // same as chart-js
                font-size: 12px;
                font-weight: normal;
                fill: var(--white);

                &.outside {
                    fill: #0f0f0f;
                }
            }
        }
    }
}
