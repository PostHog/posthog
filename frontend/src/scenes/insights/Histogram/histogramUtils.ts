import * as d3 from 'd3'
import { FunnelLayout } from 'lib/constants'
import { HistogramDatum } from 'scenes/insights/Histogram/Histogram'

export interface HistogramConfig {
    layout: FunnelLayout
    height: number
    width: number
    inner: { height: number; width: number }
    margin: { top: number; right: number; bottom: number; left: number }
    borderRadius: number
    ranges: { x: number[]; y: number[] }
    gridlineTickSize: number
    transforms: { x: string; y: string; yGrid: string }
    axisFn: { x: any; y: any }
    transitionDuration: number
    spacing: {
        btwnBins: number
        yLabel: number
        xLabel: number
        labelLineHeight: number
        barLabelPadding: number
        minBarWidth: number
    }
}

export const INITIAL_CONFIG = {
    layout: FunnelLayout.vertical,
    height: 352,
    width: 500,
    margin: { top: 20, right: 20, bottom: 20, left: 40 },
    borderRadius: 4, // same as funnel bar graph,
    transitionDuration: 200, // in ms; same as in funnel bar graph
    spacing: {
        btwnBins: 6, // spacing between bins
        yLabel: 5, // y-axis label xOffset in vertical position
        xLabel: 8, // x-axis label xOffset in horizontal position
        labelLineHeight: 1.2, // line height of wrapped label text in em's,
        barLabelPadding: 8, // padding between bar and bar label,
        minBarWidth: 90, // minimum bar width
    },
}

export const getConfig = (layout: FunnelLayout, width?: number, height?: number): HistogramConfig => {
    const _width = width || INITIAL_CONFIG.width,
        _height = height || INITIAL_CONFIG.height
    const isVertical = layout === FunnelLayout.vertical

    return {
        ...INITIAL_CONFIG,
        layout,
        height: _height,
        width: _width,
        inner: {
            height: _height - INITIAL_CONFIG.margin.bottom - INITIAL_CONFIG.margin.top,
            width: _width - INITIAL_CONFIG.margin.left - INITIAL_CONFIG.margin.right,
        },
        ranges: {
            x: isVertical
                ? [INITIAL_CONFIG.margin.left, _width - INITIAL_CONFIG.margin.right]
                : [INITIAL_CONFIG.margin.top, _height - INITIAL_CONFIG.margin.bottom],
            y: isVertical
                ? [_height - INITIAL_CONFIG.margin.bottom, INITIAL_CONFIG.margin.top]
                : [INITIAL_CONFIG.margin.left, _width - INITIAL_CONFIG.margin.right],
        },
        gridlineTickSize: isVertical
            ? _width - INITIAL_CONFIG.margin.left + INITIAL_CONFIG.spacing.yLabel - INITIAL_CONFIG.margin.right
            : _height - INITIAL_CONFIG.margin.bottom - INITIAL_CONFIG.margin.top,
        transforms: {
            x: isVertical
                ? `translate(0,${_height - INITIAL_CONFIG.margin.bottom})`
                : `translate(${INITIAL_CONFIG.margin.left},0)`,
            y: isVertical ? `translate(${INITIAL_CONFIG.margin.left},0)` : `translate(0,${INITIAL_CONFIG.margin.top})`,
            yGrid: isVertical
                ? `translate(${INITIAL_CONFIG.margin.left - INITIAL_CONFIG.spacing.yLabel},0)`
                : `translate(0,${INITIAL_CONFIG.margin.top})`,
        },
        axisFn: {
            x: isVertical ? d3.axisBottom : d3.axisLeft,
            y: isVertical ? d3.axisLeft : d3.axisTop,
        },
    }
}
// Shamelessly inspired by https://gist.github.com/skokenes/6fa266f4f50c86f77ceabcd6dfca9e42
export const createRoundedRectPath = (
    x: number,
    y: number,
    width: number,
    height: number,
    radius: number,
    position: 'top' | 'right' | 'bottom' | 'left'
): string => {
    // if empty width/height
    const isEmpty =
        (height === 0 && ['top', 'bottom'].includes(position)) || (width === 0 && ['left', 'right'].includes(position))

    const radii = {
        tl: !isEmpty && ['top', 'left'].includes(position) ? radius : 0,
        tr: !isEmpty && ['top', 'right'].includes(position) ? radius : 0,
        bl: !isEmpty && ['bottom', 'left'].includes(position) ? radius : 0,
        br: !isEmpty && ['bottom', 'right'].includes(position) ? radius : 0,
    }

    return (
        // Move to position, offset by radius in x direction
        'M' +
        (x + radii.tl + ',' + y) +
        // Draw a horizontal line to the top right curve start
        'h' +
        (width - radii.tl - radii.tr) +
        // Draw the top right corner curve
        'a' +
        radii.tr +
        ',' +
        radii.tr +
        ' 0 0 1 ' +
        radii.tr +
        ',' +
        radii.tr +
        // Draw a vertical line to the bottom right corner
        'v' +
        (height - radii.tr - radii.br) +
        // Draw the bottom right corner curve
        'a' +
        radii.br +
        ',' +
        radii.br +
        ' 0 0 1 ' +
        -radii.br +
        ',' +
        radii.br +
        // Draw a horizontal line to the bottom left corner
        'h' +
        (radii.br + radii.bl - width) +
        // Draw the bottom left corner
        'a' +
        -radii.bl +
        ',' +
        -radii.bl +
        ' 0 0 1 ' +
        -radii.bl +
        ',' +
        -radii.bl +
        // Draw a vertical line to the top left corner
        'v' +
        (radii.bl + radii.tl - height) +
        // Draw the top left corner
        'a' +
        radii.tl +
        ',' +
        -radii.tl +
        ' 0 0 1 ' +
        radii.tl +
        ',' +
        -radii.tl +
        // Close the shape
        'z'
    )
}

export interface D3HistogramDatum extends HistogramDatum {
    labelWidth?: number
    labelHeight?: number
    shouldShowInBar?: boolean
}
