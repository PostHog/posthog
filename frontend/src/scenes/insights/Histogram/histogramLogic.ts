import { kea } from 'kea'
import { getConfig, HistogramConfig } from 'scenes/insights/Histogram/histogramUtils'
import type { histogramLogicType } from './histogramLogicType'
import { FunnelLayout } from 'lib/constants'

export const histogramLogic = kea<histogramLogicType>({
    path: ['scenes', 'insights', 'Histogram', 'histogramLogic'],
    actions: {
        setConfig: (config: HistogramConfig) => ({ config }),
    },
    reducers: {
        config: [
            getConfig(FunnelLayout.vertical),
            {
                setConfig: (state, { config }) => ({ ...state, ...config }),
            },
        ],
    },
})
