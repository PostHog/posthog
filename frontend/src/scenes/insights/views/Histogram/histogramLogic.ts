import { kea, path, actions, reducers } from 'kea'
import { getConfig, HistogramConfig } from 'scenes/insights/views/Histogram/histogramUtils'
import type { histogramLogicType } from './histogramLogicType'
import { FunnelLayout } from 'lib/constants'

export const histogramLogic = kea<histogramLogicType>([
    path(['scenes', 'insights', 'Histogram', 'histogramLogic']),
    actions({
        setConfig: (config: HistogramConfig) => ({ config }),
    }),
    reducers({
        config: [
            getConfig(FunnelLayout.vertical),
            {
                setConfig: (state, { config }) => ({ ...state, ...config }),
            },
        ],
    }),
])
