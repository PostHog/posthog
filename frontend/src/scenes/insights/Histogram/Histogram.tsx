import React, { useEffect } from 'react'
import * as d3 from 'd3'
import { D3Selector, D3Transition, useD3 } from 'lib/hooks/useD3'
import { FunnelLayout } from 'lib/constants'
import { createRoundedRectPath, D3HistogramDatum, getConfig, INITIAL_CONFIG } from './histogramUtils'
import { animate, getOrCreateEl, wrap } from 'lib/utils/d3Utils'

import './Histogram.scss'
import { useActions, useValues } from 'kea'
import { histogramLogic } from 'scenes/insights/Histogram/histogramLogic'

export interface HistogramDatum {
    id: string | number
    bin0: number
    bin1: number
    count: number
    label: string | number
}

interface HistogramProps {
    data: HistogramDatum[]
    layout?: FunnelLayout
    isAnimated?: boolean
    isDashboardItem?: boolean
    width?: number
    height?: number
    formatXTickLabel?: (value: number) => number | string
    formatYTickLabel?: (value: number) => number
}

export function Histogram({
    data,
    layout = FunnelLayout.vertical,
    width = INITIAL_CONFIG.width,
    height = INITIAL_CONFIG.height,
    isAnimated = false,
    isDashboardItem = false,
    formatXTickLabel = (value: number) => value,
    formatYTickLabel = (value: number) => value,
}: HistogramProps): JSX.Element {
    const { config } = useValues(histogramLogic)
    const { setConfig } = useActions(histogramLogic)
    const isEmpty = data.length === 0 || d3.sum(data.map((d) => d.count)) === 0

    // Initialize x-axis and y-axis scales
    const xMin = data?.[0]?.bin0 || 0
    const xMax = data?.[data.length - 1]?.bin1 || 1
    const x = d3.scaleLinear().domain([xMin, xMax]).range(config.ranges.x).nice()
    const xAxis = config.axisFn
        .x(x)
        .tickValues([...data.map((d) => d.bin0), xMax])
        // v === -2 || v === -1 represent bins that catch grouped outliers.
        // TODO: (-2, -1) are temporary placeholders for (-inf, +inf) and should be changed when backend specs are finalized
        .tickFormat((v: number) => {
            const label = formatXTickLabel(v)
            if (v === -2) {
                return `<${label}`
            }
            if (v === -1) {
                return `>=${label}`
            }
            return label
        })

    // y-axis scale
    const yMax = d3.max(data, (d: HistogramDatum) => d.count) as number
    const y = d3.scaleLinear().domain([0, yMax]).range(config.ranges.y).nice()
    const yAxis = config.axisFn
        .y(y)
        .tickValues(y.ticks().filter((tick) => Number.isInteger(tick)))
        .tickSize(0)
        .tickFormat((v: number) => {
            const count = formatYTickLabel(v)
            // SI-prefix with trailing zeroes trimmed off
            return d3.format('~s')(count)
        })

    // y-axis gridline scale
    const yAxisGrid = config.axisFn.y(y).tickSize(-config.gridlineTickSize).tickFormat('').ticks(y.ticks().length)

    // Update config to new values if dimensions change
    useEffect(() => {
        const minWidth = Math.max(
            width,
            data.length * (config.spacing.minBarWidth + config.spacing.btwnBins) +
                config.margin.left +
                config.margin.right
        )
        setConfig(getConfig(layout, isDashboardItem ? width : minWidth, height))
    }, [data.length, layout, width, height])

    const ref = useD3(
        (container) => {
            const isVertical = config.layout === FunnelLayout.vertical

            const renderCanvas = (parentNode: D3Selector): D3Selector => {
                // Update config to reflect dimension changes
                x.range(config.ranges.x)
                y.range(config.ranges.y)
                yAxisGrid.tickSize(-config.gridlineTickSize)

                // Get or create svg > g
                const _svg = getOrCreateEl(parentNode, 'svg > g', () =>
                    parentNode
                        .append('svg:svg')
                        .attr('viewBox', `0 0 ${config.inner.width} ${config.inner.height}`)
                        .attr('width', '100%')
                        .append('svg:g')
                        .classed(config.layout, true)
                )
                // update dimensions
                parentNode.select('svg').attr('viewBox', `0 0 ${config.width} ${config.height}`)

                // if class doesn't exist on svg>g, layout has changed. after we learn this, reset
                // the layout
                const layoutChanged = !_svg.classed(config.layout)
                _svg.attr('class', null).classed(config.layout, true)

                // if layout changes, redraw axes from scratch
                if (layoutChanged) {
                    _svg.selectAll('#x-axis,#y-axis,#y-gridlines').remove()
                }

                // x-axis
                const _xAxis = getOrCreateEl(_svg, 'g#x-axis', () =>
                    _svg.append('svg:g').attr('id', 'x-axis').attr('transform', config.transforms.x)
                )
                _xAxis.call(animate, !layoutChanged ? config.transitionDuration : 0, isAnimated, (it: D3Transition) =>
                    it.call(xAxis).attr('transform', config.transforms.x)
                )
                const binWidth = x(data?.[0]?.bin1 || data?.[data.length - 1]?.bin1 || 1) - x(data?.[0]?.bin0 || 0)
                _xAxis
                    .selectAll('.tick text')
                    .call(
                        wrap,
                        isVertical ? binWidth : config.margin.left,
                        config.spacing.labelLineHeight,
                        isVertical,
                        config.spacing.xLabel
                    )

                // Don't draw y-axis or y-gridline if the data is empty
                if (!isEmpty) {
                    // y-axis
                    const _yAxis = getOrCreateEl(_svg, 'g#y-axis', () =>
                        _svg.append('svg:g').attr('id', 'y-axis').attr('transform', config.transforms.y)
                    )
                    _yAxis.call(
                        animate,
                        !layoutChanged ? config.transitionDuration : 0,
                        isAnimated,
                        (it: D3Transition) =>
                            it
                                .call(yAxis)
                                .attr('transform', config.transforms.y)
                                .call((g) =>
                                    g.selectAll('.tick text').attr('dx', isVertical ? `-${config.spacing.yLabel}` : 0)
                                )
                    )

                    // y-gridlines
                    const _yGridlines = getOrCreateEl(_svg, 'g#y-gridlines', () =>
                        _svg.append('svg:g').attr('id', 'y-gridlines').attr('transform', config.transforms.yGrid)
                    )
                    _yGridlines.call(
                        animate,
                        !layoutChanged ? config.transitionDuration : 0,
                        isAnimated,
                        (it: D3Transition) => it.call(yAxisGrid).attr('transform', config.transforms.yGrid)
                    )
                }

                const d3Data = data as D3HistogramDatum[]

                // bars
                const _bars = getOrCreateEl(_svg, 'g#bars', () => _svg.append('svg:g').attr('id', 'bars'))
                _bars
                    .selectAll('path')
                    .data(d3Data)
                    .join('path')
                    .call(animate, config.transitionDuration, isAnimated, (it: D3Transition) => {
                        return it.attr('d', (d: HistogramDatum) => {
                            if (!isVertical) {
                                // is horizontal
                                return createRoundedRectPath(
                                    y(0),
                                    x(d.bin0) + config.spacing.btwnBins / 2,
                                    y(d.count) - y(0),
                                    Math.max(0, x(d.bin1) - x(d.bin0) - config.spacing.btwnBins),
                                    config.borderRadius,
                                    'right'
                                )
                            }
                            return createRoundedRectPath(
                                x(d.bin0) + config.spacing.btwnBins / 2,
                                y(d.count),
                                Math.max(0, x(d.bin1) - x(d.bin0) - config.spacing.btwnBins),
                                y(0) - y(d.count),
                                config.borderRadius,
                                'top'
                            )
                        })
                    })

                // Always move bar above everything else
                _svg.node().appendChild(_bars.node())

                // text labels
                const _labels = getOrCreateEl(_svg, 'g#labels', () => _svg.append('svg:g').attr('id', 'labels'))
                _labels
                    .selectAll('text')
                    .data(d3Data)
                    .join('text')
                    .text((d) => d.label)
                    .classed('bar-label', true)
                    .each(function (this: any, d) {
                        const { width: labelWidth, height: labelHeight } = this.getBBox()
                        d.labelWidth = labelWidth
                        d.labelHeight = labelHeight
                        d.shouldShowInBar = false
                    })
                    .attr('x', (d) => {
                        if (!isVertical) {
                            const labelWidth = (d.labelWidth || 0) + 2 * config.spacing.barLabelPadding
                            const shouldShowInBar = labelWidth <= y(d.count) - y(0)
                            const labelDx = shouldShowInBar
                                ? -(labelWidth - config.spacing.barLabelPadding)
                                : config.spacing.barLabelPadding
                            d.shouldShowInBar = shouldShowInBar
                            return y(d.count) + labelDx
                        }
                        // x + bin width + dx + dy
                        return x(d.bin0) + binWidth / 2 - (d.labelWidth || 0) / 2
                    })
                    .attr('y', (d) => {
                        if (!isVertical) {
                            return x(d.bin0) + binWidth / 2
                        }
                        // determine if label should be in the bar or above it.
                        const labelHeight = (d.labelHeight || 0) + 2 * config.spacing.barLabelPadding
                        const shouldShowInBar = labelHeight <= y(0) - y(d.count)
                        const labelDy = shouldShowInBar
                            ? labelHeight - config.spacing.barLabelPadding
                            : -config.spacing.barLabelPadding
                        d.shouldShowInBar = shouldShowInBar
                        return y(d.count) + labelDy
                    })
                    .classed('outside', (d) => !d.shouldShowInBar)

                // Always move labels to top
                _svg.node().appendChild(_labels.node())

                return _svg
            }

            renderCanvas(container)
        },
        [data, config]
    )

    /* minWidth required to enforce d3's width calculations on the div wrapping the svg
     so that scrolling horizontally works */
    return <div className="histogram-container" ref={ref} style={{ minWidth: config.width }} />
}
