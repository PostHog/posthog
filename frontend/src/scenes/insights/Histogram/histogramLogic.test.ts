import { histogramLogic } from './histogramLogic'
import { initKeaTests } from '~/test/init'
import { getConfig } from 'scenes/insights/Histogram/histogramUtils'
import { FunnelLayout } from 'lib/constants'

describe('histogramLogic', () => {
    let logic: ReturnType<typeof histogramLogic.build>

    beforeEach(() => {
        initKeaTests()
        logic = histogramLogic()
        logic.mount()
    })

    describe('values', () => {
        it('has proper defaults', () => {
            expect(logic.values).toMatchSnapshot()
        })
    })

    describe('reducers', () => {
        describe('config', () => {
            it('is set via setConfig', async () => {
                const newConfig = getConfig(FunnelLayout.vertical, 1000, 3000)
                expect(logic.values.config).toEqual(getConfig(FunnelLayout.vertical))
                logic.actions.setConfig(newConfig)
                expect(logic.values.config).toEqual(newConfig)
            })
        })
    })
})
