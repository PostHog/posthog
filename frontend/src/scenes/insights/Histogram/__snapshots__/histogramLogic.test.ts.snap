// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`histogramLogic values has proper defaults 1`] = `
Object {
  "config": Object {
    "axisFn": Object {
      "x": [Function],
      "y": [Function],
    },
    "borderRadius": 4,
    "gridlineTickSize": 445,
    "height": 500,
    "inner": Object {
      "height": 460,
      "width": 440,
    },
    "layout": "vertical",
    "margin": Object {
      "bottom": 20,
      "left": 40,
      "right": 20,
      "top": 20,
    },
    "ranges": Object {
      "x": Array [
        40,
        480,
      ],
      "y": Array [
        480,
        20,
      ],
    },
    "spacing": Object {
      "barLabelPadding": 8,
      "btwnBins": 6,
      "labelLineHeight": 1.2,
      "minBarWidth": 90,
      "xLabel": 8,
      "yLabel": 5,
    },
    "transforms": Object {
      "x": "translate(0,480)",
      "y": "translate(40,0)",
      "yGrid": "translate(35,0)",
    },
    "transitionDuration": 200,
    "width": 500,
  },
}
`;
